core/src/main/scala/org/apache/spark/sql/delta/GeneratedColumn.scala

```diff
@@ -362,6 +362,31 @@ object GeneratedColumn extends DeltaLogging with AnalysisHelper {
                   createExpr(name)(SubstringPartitionExpr(partColName, pos, len))
                 case ExtractBaseColumn(name, _) =>
                   createExpr(name)(IdentityPartitionExpr(partColName))
+                case TruncDate(ExtractBaseColumn(name, DateType), StringLiteral(format)) =>
+                  format toUpperCase Locale.ROOT match {
+                    case DATE_FORMAT_WEEK =>
+                      createExpr(name)(TruncDatePartitionExpr(partColName, DATE_FORMAT_WEEK))
+                    case DATE_FORMAT_QUARTER =>
+                      createExpr(name)(TruncDatePartitionExpr(partColName, DATE_FORMAT_QUARTER))
+                    case DATE_FORMAT_YEAR | DATE_FORMAT_YEAR_2 | DATE_FORMAT_YEAR_3 =>
+                      createExpr(name)(TruncDatePartitionExpr(partColName, DATE_FORMAT_YEAR))
+                    case DATE_FORMAT_MONTH | DATE_FORMAT_MONTH_2 | DATE_FORMAT_MONTH_3 =>
+                      createExpr(name)(TruncDatePartitionExpr(partColName, DATE_FORMAT_MONTH))
+                    case _ => None
+                  }
+                case TruncDate(Cast(
+                ExtractBaseColumn(name, _), DateType, _, _), StringLiteral(format)) =>
```

Comment:
That will not consider the casting from string to date. Should I split them into two cases, one for TimeStamp and the other for String?

Reply:
I just figure out that I can let them this way, to avoid having one more case, `TruncDate(Cast(ExtractBaseColumn(name, TimestampType | StringType), DateType, _, _), StringLiteral(format))`